Support multiclass classification

wxchan opened the pull request:
to solve #43 

I'd like to open this PR to let more people see and help build or test on this feature. [Lastest version](https://github.com/wxchan/LightGBM/tree/dev).

You need to set `objective = multiclass`, `metric = multi_logloss(or multi_error)` and `num_class =` in train.conf, `num_class =` in predict.conf is not needed any more. There is an example in example/multiclass_classification.

guolinke commented:
@wxchan Thanks for your nice work.
However, this seems will have conflict with my PR for the prepare of c_api. I refine some interface on boosting.h and metric.h. 
Maybe you can continue after I finish the changes for interface refine.
Thanks again.

chivee commented:
Hi, @wxchan , you can rebase master for the new c_api.

BTW, could you please also give some test example for this PR?

wxchan commented:
@guolinke I fixed the conflicts according to your new commits.

@chivee I already added an example in examples/multiclass_classification. You can run it and use compare.py [here](https://github.com/wxchan/LightGBM/commit/3449b0c71ae3cf780383a7ae513b75624e7463b2) to compare its performance with xgboost.

wxchan commented:
A small bug when save the model. I think [this line](https://github.com/Microsoft/LightGBM/blob/master/src/application/application.cpp#L251) is not necessary. Because you already pop_back the trees if meets early stopping. The remaining trees will save at next SaveModelToFile(). After my commit, some trees will save twice, can't figure out why.

Btw, is it necessary to delete back() before pop_back()? I thought pop_back() ([see docs](http://www.cplusplus.com/reference/vector/vector/pop_back/)) will delete it anyway.

guolinke commented:
@wxchan I think pop_back() cannot free the resource of a pointer. 
refer to: http://stackoverflow.com/questions/10696124/vectorclear-pop-back-wont-run-destructor
https://www.daniweb.com/programming/software-development/threads/170492/does-std-vector-s-pop-back-free-the-memory
http://stackoverflow.com/questions/10696124/vectorclear-pop-back-wont-run-destructor

guolinke commented:
@wxchan  to avoid save repeated model, you need to change [this](https://github.com/wxchan/LightGBM/blob/dev/src/boosting/gbdt.cpp#L328-L333) as well.

wxchan commented:
@guolinke Yes, I know, but I can't figure out how. (I already changed ` early_stopping_round_` to `early_stopping_round_ * num_class_`, but it's not where the problem is)

Actually, I am not sure your origin code is right. For example, set early_stopping_round_ = 10, and assume when iter = 50, it meets early stopping criteria. In this situation, at iter 50, you already write first 40 trees to file. Then because of you pop last 10 trees in TrainOneIter(), the models_.size() = 40. So, rest = 40 - 10 = 30. Then at line 336, saved_model_size_ = rest = 30. So, in line 341-348, 31-40 trees are saved again. 

You can try to remove the SaveModelToFile() line I mentioned above and see it works. At least the result seems right.

guolinke commented:
@wxchan Thanks, you are right. I will fix this directly, then merge your PR after some other reviews.

guolinke commented:
@wxchan I fix this by change this https://github.com/Microsoft/LightGBM/blob/master/src/boosting/gbdt.cpp#L321 .
Use "iter_" is not so elegant.

guolinke commented on src/metric/multiclass_metric.hpp:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  void Init(const char* test_name, const Metadata& metadata, data_size_t num_data) override {
+    name_ = test_name;
```
name_ = test_name+metric_name..

guolinke commented on src/metric/multiclass_metric.hpp:
```diff
@@ -0,0 +1,135 @@
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        std::vector<score_t> rec(num_class_);
+        for (int k = 0; k < num_class_; ++k) {
+          rec[k] = score[k * num_data_ + i];
```
don't need softmax before calculate logloss?

guolinke commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,98 @@
+        std::vector<score_t> rec(num_class_);
+        int ilabel = static_cast<int>(label_[i]);  
+        if (ilabel < 0 || ilabel >= num_class_) {
+            Log::Fatal("Label must be in [0, num_class)");
```
This checking should be done in "Init".

guolinke commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,98 @@
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        std::vector<score_t> rec(num_class_);
+        int ilabel = static_cast<int>(label_[i]);  
```
will cast to int every time. Maybe using an int array to cache this is better

guolinke commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  float GetSigmoid() const override {
+    return -1.0;
```
-1.0f

guolinke commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,98 @@
+  int* label_val_;
+  /*! \brief Weights for multiclass labels */
+  score_t* label_weights_;
+  /*! \brief Weights for data */
```
remove unused variables. is_unbalance_, label_val_, label_weights_

guolinke commented on src/boosting/score_updater.hpp:
```diff
@@ -18,12 +18,12 @@ class ScoreUpdater {
-  explicit ScoreUpdater(const Dataset* data)
-    :data_(data) {
+  explicit ScoreUpdater(const Dataset* data, int num_class)
+    :data_(data), num_class_(num_class) {
```
Should we really need to pass "num_class" to score_updater?
I think we can add an parameter "offset" in every "AddScore". 
And pass "num_data_*num_class_" as size in constructor

guolinke commented on src/boosting/gbdt.h:
```diff
@@ -180,6 +199,10 @@ class GBDT: public Boosting {
+  /*! \brief Number of classes */
+  int num_class_;
+  /*! \brief Current class */
+  int curr_class_;
```
really use this?  Or add "curr_class" as parameter in these functions who use "curr_class"?

wxchan replied:
curr_class_ is used in Bagging(), UpdateScoreOutOfBag(), TrainOneIter() and UpdateScore(). It indicates current tree_learner we are using. It works just like iter_.

guolinke replied:
I know this. My point is should we really need to use "curr_class_" as a member variable?
why not pass "curr_class" as the argument of these functions?

guolinke commented on src/boosting/gbdt.cpp:
```diff
@@ -471,4 +496,4 @@ std::vector<int> GBDT::PredictLeafIndex(const float* value, int num_used_model)
 }
 
-}  // namespace LightGBM
+}  // namespace LightGBM
```
an new line ...

xuehui1991 commented on include/LightGBM/utils/common.h:
```diff
@@ -336,6 +336,22 @@ static inline int64_t Pow2RoundUp(int64_t x) {
   return 0;
 }
 
+inline void Softmax(std::vector<float>* p_rec) {
```
I wonder what is the abbreviation of "p_rec"? can you rename this parameter?

xuehui1991 commented on include/LightGBM/utils/common.h:
```diff
@@ -336,6 +336,22 @@ static inline int64_t Pow2RoundUp(int64_t x) {
 
+inline void Softmax(std::vector<float>* p_rec) {
+  std::vector<float> &rec = *p_rec;
+  float wmax = rec[0];
```
what's meaning of "wmax"?

wxchan replied:
@xuehui1991 to be honest, I don't know. I copied it from [xgboost](https://github.com/dmlc/xgboost/blob/master/src/common/math.h#L30). The result is same with or without it.

xuehui1991 commented on src/application/predictor.hpp:
```diff
@@ -120,7 +133,20 @@ class Predictor {
+    if (num_class_ > 1) {
+      predict_fun = [this](const std::vector<std::pair<int, float>>& features){
+        std::vector<float> prediction = PredictMulticlassOneLine(features);
+        std::stringstream result_ss;
```
rename to "result_stream_buf"?

xuehui1991 commented on src/boosting/gbdt.cpp:
```diff
@@ -146,52 +152,60 @@ void GBDT::Bagging(int iter) {
-  // update score
-  UpdateScore(new_tree);
-  UpdateScoreOutOfBag(new_tree);
+    
```
why add a blank line here?

guolinke commented on src/io/config.cpp:
```diff
@@ -45,7 +45,27 @@ void OverallConfig::Set(const std::unordered_map<std::string, std::string>& para
+      }      
+  }
+  for (std::string metric_type : metric_types){
+        bool metric_type_multiclass = ( metric_type == "multi_logloss" || metric_type == "multi_error");
```
It is better to use  metric_type == std::string("multi_logloss" )

chivee commented on src/io/config.cpp:
```diff
@@ -45,7 +45,27 @@ void OverallConfig::Set(const std::unordered_map<std::string, std::string>& para
-
+  
+  bool objective_type_multiclass = (objective_type == "multiclass");
+  int num_class_test = 1;
```
please using assert for runtime logic check

guolinke replied:
@wxchan you can direct use "GetInt(params, "num_class", &num_class);" then test the conflict in the function "CheckParamConflict"

xuehui1991 commented on src/metric/multiclass_metric.hpp:
```diff
@@ -0,0 +1,138 @@
+      }
+    }
+    score_t loss = sum_loss / sum_weights_;
+    return std::vector<score_t>(1, loss);;
```
why there are two "；"？

xuehui1991 commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,92 @@
+
+namespace LightGBM {
+/*!
+* \brief Objective funtion for multiclass classification
```
function?

xuehui1991 commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,92 @@
+    num_data_ = num_data;
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+    ilabel_ = new int[num_data_];
```
can you change the name "ilabel_"?

wxchan commented on src/application/application.cpp:
```diff
@@ -269,6 +269,7 @@ void Application::Predict() {
 void Application::InitPredict() {
   boosting_ =
     Boosting::CreateBoosting(config_.io_config.input_model.c_str());
+  boosting_ -> SetNumberOfClass(config_.boosting_config->num_class);
```
@guolinke it's a great idea. We don't need num_class in predict.conf then.

guolinke commented on src/application/predictor.hpp:
```diff
@@ -33,6 +33,7 @@ class Predictor {
       num_used_model_(num_used_model) {
     boosting_ = boosting;
     num_features_ = boosting_->MaxFeatureIdx() + 1;
+    num_class_ = boosting_->NumberOfClass();
```
"num_class_" is not used in predictor?

guolinke commented on src/objective/multiclass_objective.hpp:
```diff
@@ -0,0 +1,95 @@
+          } else {
+            gradients[k * num_data_ + i] = p * weights_[i];
+          }
+          hessians[k * num_data_ + i] = 2.0f * p * (1.0f - p) * weights_[i];
```
@wxchan Did you know why here need to multiply ```2.0``` ?  From my calculation, it didn't has this factor.

wxchan replied:
I referred it to xgboost a little: https://github.com/dmlc/xgboost/blob/master/src/objective/multiclass_obj.cc#L66

I am not sure why it needs to multiply 2 now, seem necessary.

Laurae2 replied:
In Newton method, the second derivative for it is: `p*(1-p)` but it is only for the binary case.

For softmax it seems it comes from another approximation method, on this paper they use 2 on the `B^-1` part for the upper bounding computation so the formula can be used as the hessian instead: http://www.ism.ac.jp/editsec/aism/pdf/044_1_0197.pdf (formula c and d, last page)

That `2` does not show up in the binary case (Newton method).

guolinke replied:
Thanks!

guolinke replied:
I found this from sklearn.gbm: https://github.com/scikit-learn/scikit-learn/blob/919b4a8fbdf0e313fb702bd083abb31d67f7d8f9/sklearn/ensemble/gradient_boosting.py#L566-L584 
It seems it multiplies the `(num_class-1)/num_class` .

Laurae2 replied:
In matlab it seems `1/2` for softmax: https://fr.mathworks.com/matlabcentral/fileexchange/55863-logistic-regression-for-classification?focused=6032429&tab=function

Newton Raphson / IRLS: `f(x) = f(x_t) + df(x - x_t)/dx_t + (1/2)*(x - x_t)^T * d²f(x_t * (x - x_t))/(dxdxT) + o(||x - x_t||²)`.

As long as the number is not larger than the maximum allowed (`2` apparently on the paper I linked before) it should converge.
